opm/simulators/wells/WellAndGroupStates.hpp

```diff
@@ -0,0 +1,564 @@
+        for (int w = 0; w < nw; ++w) {
+            well_names_[w] = wells[w].name();
+            initSingleWell(cell_pressures, schedule, wells[w], report_step, phase_usage, well_perf_data_[w], summary_state, well_states_[w]);
+        }
+
+        // For wells present in a previous state, override initialization where appropriate.
+        if (prev_state) {
+            for (int w = 0; w < nw; ++w) {
+                well_names_[w] = wells[w].name();
+                auto it = std::find(prev_state->well_names_.begin(), prev_state->well_names_.end(), well_names_[w]);
+                if (it != prev_state->well_names_.end()) {
+                    const int w_in_prev = it - prev_state->well_names_.end();
+                    initSingleWellFromPrevious(prev_state->well_states_[w_in_prev], well_states_[w]);
+                }
+            }
+        }
+
+        // Group states remain default-initialized.
+    }
+
+    data::Wells report(const PhaseUsage& phase_usage, const int* globalCellIdxMap) const
+    {
+        data::Wells dwells;
+        const int nw = well_states_.size();
+        for (int w = 0; w < nw; ++w) {
+            if (well_states_[w].status == Well::Status::SHUT) {
+                continue;
+            }
+            auto& dwell = dwells[well_names_[w]];
+            reportSingleWell(phase_usage, globalCellIdxMap, well_perf_data_[w], well_states_[w], dwell);
+        }
+        return dwells;
+    }
+
+    using WellState = SingleWellState<NumActivePhases>;
+    using GroupState = SingleGroupState<NumActivePhases>;
+
+    std::vector<WellState>& wellStates()
+    {
+        return well_states_;
+    }
+    const std::vector<WellState>& wellStates() const
+    {
+        return well_states_;
+    }
+
+    std::map<std::string, GroupState>& groupStates()
+    {
+        return group_states_;
+    }
+    const std::map<std::string, GroupState>& groupStates() const
+    {
+        return group_states_;
+    }
+
+
+private:
+
+    // -----------  Data members  -----------
+
+    std::vector<WellState> well_states_;
+    std::vector<std::string> well_names_;
+    std::map<std::string, GroupState> group_states_;
+
+    std::vector<std::vector<PerforationData>> well_perf_data_;
+
+
+    // -----------  Private functions  -----------
+
+    template <unsigned long N>
+    static double sum(const std::array<double, N>& x)
+    {
+        return std::accumulate(x.begin(), x.end(), 0.0);
+    }
+
+    static void initSingleWell(const std::vector<double>& cell_pressures,
+                               const Schedule& schedule,
+                               const Well& well,
+                               const int report_step,
+                               const PhaseUsage& phase_usage,
+                               const std::vector<PerforationData>& perf_data,
+                               const SummaryState& summary_state,
+                               SingleWellState<NumActivePhases>& wstate)
```

Comment:
I guess that is a matter of taste - but I would prefer to have the `self` argument - i.e. `SingleWellState<>& wstate` as the *first* argument.

Reply:
It is not a *self* argument, since the class here is the WellAndGroupStates, so I put the output argument last as is customary. Is there some rule-of-thumb I am missing?

Reply:
Well - loosely speaking it is akin to a `self` argument because that is the object being manipulated in this function - maybe it could be a method in the `SingleWellState`class instead?

Reply:
> maybe it could be a method in the `SingleWellState`class instead?

It could, but I preferred to keep all the initialization code together in the same class and file. It also makes the SingleWellState class depend on much fewer extra headers.  I suggest we leave it like that for now since it is likely to change anyway as the well model takes responsibility for init?

Reply:
> I suggest we leave it like that for now
Sure